lib/Sockets.js

```diff
@@ -0,0 +1,269 @@
+      bot.findConversation({
+        user: message.user,
+        channel: message.channel
+      }, function(convo) {
+          if (convo) {
+            if (bot.ws) {
+              // replace the websocket connection
+              convo.task.bot.ws = bot.ws;
+              convo.task.bot.connected = true;
+              if (message.type == 'hello' || message.type == 'welcome_back') {
+                message.type = 'reconnect';
+              }
+            } else {
+
+              // replace the reply channel in the active conversation
+              // this is the one that gets used to send the actual reply
+              convo.task.bot.http_response = bot.http_response;
+          }
+        }
+        next();
+      });
+    });
+
+    controller.middleware.categorize.use(function(bot, message, next) {
+
+        if (message.type == 'message') {
+            message.type = 'message_received';
+        }
+
+        next();
+
+    });
+
+    // simple message clone because its already in the right format!
+    controller.middleware.format.use(function(bot, message, platform_message, next) {
+
+        for (var key in message) {
+            platform_message[key] = message[key];
+        }
+        if (!platform_message.type) {
+          platform_message.type = 'message';
+        }
+        next();
+
+    });
+
+
+    controller.defineBot(function(botkit, config) {
+        var bot = {
+            type: 'socket',
+            botkit: botkit,
+            config: config || {},
+            utterances: botkit.utterances,
+        };
+
+        bot.startConversation = function(message, cb) {
+            botkit.startConversation(this, message, cb);
+        };
+
+        bot.createConversation = function(message, cb) {
+            botkit.createConversation(this, message, cb);
+        };
+
+        bot.send = function(message, cb) {
+            if (bot.connected || !bot.ws) {
+                try {
+                    if (bot.ws) {
+                      bot.ws.send(JSON.stringify(message));
+                    } else {
+                      bot.http_response.json(message);
+                    }
+                } catch (err) {
+                    console.error('ERROR SENDING', err);
+                    if (cb) return cb(err, message);
+                }
+                if (cb) cb(null, message);
+
+            } else {
+
+                // console.log('not connected. wait to resend');
+                setTimeout(function() {
+                    bot.send(message, cb);
+                }, 3000);
+            }
+        };
+
+        bot.startTyping = function() {
+            if (bot.connected) {
+                bot.ws.send(JSON.stringify({
+                    type: 'typing',
+                }));
+            }
+        };
+
+        bot.typingDelay = function(message) {
+
+            return new Promise(function(resolve, reject) {
```

Comment:
There is no `require("promise")` at the top  of this file, so this code is using Ecmascript6 Promise object.

I've never used it, so I don't know if it is important, but botkit has dependency on `promise` library: https://github.com/howdyai/botkit/blob/dev/webchannel/package.json#L24 which is used in https://github.com/howdyai/botkit/blob/dev/webchannel/lib/Studio.js#L2 and I think that it would be the best to use the same promise library in Sockets.js